app/maptools/recordingmaptool.cpp

```diff
@@ -406,3 +401,94 @@ void RecordingMapTool::setHandles( const QgsGeometry &newHandles )
   mHandles = newHandles;
   emit handlesChanged( mHandles );
 }
+
+const QString &RecordingMapTool::state() const
+{
+  return mState;
+}
+
+void RecordingMapTool::setState( const QString &newState )
+{
+  if ( mState == newState )
+    return;
+  mState = newState;
+  emit stateChanged( mState );
+}
+
+QgsVertexId &RecordingMapTool::clickedVertexId()
+{
+  return mClickedVertexId;
+}
+
+void RecordingMapTool::setClickedVertexId( QgsVertexId newId )
+{
+  if ( mClickedVertexId == newId )
+    return;
+  mClickedVertexId.part = newId.part;
+  mClickedVertexId.ring = newId.ring;
+  mClickedVertexId.vertex = newId.vertex;
+  emit clickedVertexIdChanged( mClickedVertexId );
+}
+
+void RecordingMapTool::lookForVertex( const QPointF &clickedPoint, double searchRadius )
+{
+  double minDistance = std::numeric_limits<double>::max();
+  double currentDistance = 0;
+  QgsPoint point;
+  QgsVertexId vertexId;
+
+  QgsPoint pnt = mapSettings()->screenToCoordinate( clickedPoint );
+
+  if ( mInitialGeometry.isEmpty() )
+  {
+    setClickedVertexId( vertexId );
+    return;
+  }
+
+  for ( QPair<QgsVertexId, QgsPoint> pair : mVertexIds )
+  {
+    currentDistance = QgsGeometryUtils::sqrDistance2D( pnt, pair.second );
+    if ( currentDistance <= minDistance && currentDistance <= searchRadius )
+    {
+      minDistance = currentDistance;
+      vertexId.part = pair.first.part;
+      vertexId.ring = pair.first.ring;
+      vertexId.vertex = pair.first.vertex;
+    }
+  }
+
+  setClickedVertexId( vertexId );
+}
+
+void RecordingMapTool::removeVertex( QgsVertexId id )
+{
+  QgsGeometry geometry;
+  QgsAbstractGeometry *geom = mInitialGeometry.get()->clone();
+
+  if ( geom->deleteVertex( id ) )
+    geometry.set( geom );
+
+  setRecordedGeometry( geometry );
+}
+
+void RecordingMapTool::insertVertex( QgsVertexId id, const QgsPoint &point )
+{
+  QgsGeometry geometry;
+  QgsAbstractGeometry *geom = mInitialGeometry.get()->clone();
+
+  if ( geom->insertVertex( id, point ) )
+    geometry.set( geom );
+
+  setRecordedGeometry( geometry );
+}
+
+void RecordingMapTool::updateVertex( QgsVertexId id, const QgsPoint &point )
+{
+  QgsGeometry geometry;
+  QgsAbstractGeometry *geom = mInitialGeometry.get()->clone();
```

Comment:
Again, we probably want to take `mRecordedGeometry` instead of `mInitialGeometry`